File: Programmers/Level 2/Level2_롤케이크자르기.py
topping = [1, 2, 1, 3, 1, 4, 1, 2]

def toppingTypeCount(sliceCake):
    toppingCount = 0
    for i in range(1, len(sliceCake)):
        if sliceCake[i] != 0:
            toppingCount += 1
    return toppingCount

def solution(topping):
    answer = 0

    left_cake = [0] * (max(topping) + 1)
    right_cake = [0] * (max(topping) + 1)
    for to in topping:
        right_cake[to] += 1

    left_cake[topping[0]] += 1
    right_cake[topping[0]] -= 1

    left_count = toppingTypeCount(left_cake)
    right_count = toppingTypeCount(right_cake)
    if left_count == right_count:
        answer += 1
    
    for i in range(1, len(topping)):
        left_cake[topping[i]] += 1
        right_cake[topping[i]] -= 1

        if left_cake[topping[i]] == 1:
            left_count += 1
        if right_cake[topping[i]] == 0:
            right_count -= 1

        if left_count == right_count:
            answer += 1

    return answer

print(solution(topping))



# 다른 사람 풀이 중 마음에 들었던 풀이 방식 
# collections 모듈의 Counter 클래스 사용
# 코드는 매우 만족스러울 정도로 깔끔하지만 실행해 속도를 확인해본 결과
# 가장오래걸리는 테스트케이스 기준으로 내 코드가 100ms 빨랐다
from collections import Counter

def solution_counter(topping):
    dic = Counter(topping)
    set_dic = set()
    res = 0
    for i in topping:
        dic[i] -= 1
        set_dic.add(i)
        if dic[i] == 0:
            dic.pop(i)
        if len(dic) == len(set_dic):
            res += 1
    return res

File: Programmers/Level 2/Level2_롤케이크자르기.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <queue>
#include <stack>
#include <map>
#include <set>
using namespace std;

int solution(vector<int> topping) {
    int answer = 0;

	map<int, int> leftTopping;
	map<int, int> rightTopping;

	// 초기에는 모든 토핑이 오른쪽에 있다고 가정
	for (int t : topping)
		rightTopping[t]++;

	// 왼쪽과 오른쪽의 토핑 종류 수를 비교하면서 이동
	for (int t : topping) 
	{
		// 현재 토핑을 왼쪽으로 이동
		leftTopping[t]++;
		rightTopping[t]--;

		// 오른쪽에서 해당 토핑이 더 이상 없으면 제거
		if (rightTopping[t] == 0)
			rightTopping.erase(t);

		// 왼쪽과 오른쪽의 토핑 종류 수가 같으면 공평하게 나누어진 것
		if (leftTopping.size() == rightTopping.size())
			answer++;
	}

    return answer;
}

// 1차 시도 : 성공
